src/Analyzers/CSharp/Analyzers/UseImplicitObjectCreation/CSharpUseImplicitObjectCreationDiagnosticAnalyzer.cs

```diff
@@ -93,6 +94,15 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
                     _ => null,
                 };
             }
+            else if (objectCreation.Parent.IsKind(SyntaxKind.ArrayInitializerExpression, SyntaxKind.CollectionInitializerExpression))
+            {
+                typeNode = objectCreation.Parent.Parent switch
+                {
+                    ArrayCreationExpressionSyntax arrayCreation => arrayCreation.Type.ElementType,
+                    ObjectCreationExpressionSyntax { Type: QualifiedNameSyntax { Right: GenericNameSyntax { TypeArgumentList: TypeArgumentListSyntax argumentList } } } => argumentList.Arguments.Count == 1 ? argumentList.Arguments[0] : null,
```

Comment:
this only matches if the type name is dotted.   but that doesn't seem needed.  also, it presumes the generic type parameter is the type to use, but that's not necessarily true.  It also presumes that this should only work for the single genertic type parametr case.  but consider basic things like:  `new Dictionary<string, C> { { x, new C() } }`.  

For collection initializers, we need to find the respective '`Add` method' and pulls the type out of the parameter type.  Thanks!

Reply:
> It also presumes that this should only work for the single genertic type parametr case

Yes exactly what I had in mind, but didn't have time to think of the other cases. I wanted to think of the dictionary case too.

> For collection initializers, we need to find the respective 'Add method' and pulls the type out of the parameter type.

Sounds more elegant. Will give it a go.